applications/OptimizationApplication/custom_utilities/container_variable_data_holder/container_variable_data_holder.cpp

```diff
@@ -0,0 +1,657 @@
+ContainerVariableDataHolder<TContainerType, TContainerDataIO> ContainerVariableDataHolder<TContainerType, TContainerDataIO>::operator^(const double Value) const
+{
+    ContainerVariableDataHolder<TContainerType, TContainerDataIO> result(*(this->mpModelPart));
+    result.mDataDimension = this->mDataDimension;
+    result.mData.resize(this->mData.size(), false);
+    IndexPartition<IndexType>(this->mData.size()).for_each([&](const IndexType Index) {
+        result.mData[Index] = std::pow(this->mData[Index], Value);
+    });
+
+    return result;
+}
+
+template <class TContainerType, class TContainerDataIO>
+ContainerVariableDataHolder<TContainerType, TContainerDataIO>& ContainerVariableDataHolder<TContainerType, TContainerDataIO>::operator^=(const double Value)
+{
+    IndexPartition<IndexType>(this->mData.size()).for_each([&](const IndexType Index) {
+        this->mData[Index] = std::pow(this->mData[Index], Value);
+    });
+
+    return *this;
+}
+
+template <class TContainerType, class TContainerDataIO>
+ContainerVariableDataHolder<TContainerType, TContainerDataIO>& ContainerVariableDataHolder<TContainerType, TContainerDataIO>::operator=(const ContainerVariableDataHolder<TContainerType, TContainerDataIO>& rOther)
+{
+    KRATOS_ERROR_IF(&this->GetModelPart() != &rOther.GetModelPart())
+        << "Mismatching model parts found in assignment.\n"
+        << "      Assignee data: " << *this << "\n"
+        << "      Assignor data: " << rOther << "\n";
+
+    if (this->mData.size() != rOther.mData.size()) {
+        this->mData.resize(rOther.mData.size(), false);
+    }
+
+    this->mDataDimension = rOther.mDataDimension;
+
+    IndexPartition<IndexType>(this->mData.size()).for_each([&](const IndexType Index) {
+        this->mData[Index] = rOther.mData[Index];
+    });
+
+    return *this;
+}
+
+template <class TContainerType, class TContainerDataIO>
+std::string ContainerVariableDataHolder<TContainerType, TContainerDataIO>::Info() const
+{
+    std::stringstream msg;
+
+    if constexpr(std::is_same_v<TContainerType, ModelPart::NodesContainerType>) {
+        msg << "Nodal";
+    } else if constexpr(std::is_same_v<TContainerType, ModelPart::ConditionsContainerType>) {
+        msg << "Condition";
+    } else if constexpr(std::is_same_v<TContainerType, ModelPart::ElementsContainerType>) {
+        msg << "Element";
+    }
+
+    if constexpr(std::is_same_v<TContainerDataIO, HistoricalContainerDataIO>) {
+        msg << "Historical";
+    } else if constexpr(std::is_same_v<TContainerDataIO, NonHistoricalContainerDataIO>) {
+        msg << "NonHistorical";
+    } else if constexpr(std::is_same_v<TContainerDataIO, PropertiesContainerDataIO>) {
+        msg << "Properties";
+    }
+
+    msg << BaseType::Info();
+
+    return msg.str();
+}
+
+//template instantiations
+#define INSTANTIATIE_CONTAINER_DATA_METHODS(ContainerType, ContainerDataIOType, DataType)                                                                   \
```

Comment:
I know these macro names are already long, but I still recommend slapping `KRATOS_` in the front to avoid any surprising macro name clashes.

Reply:
done :)